### PR TITLE
merge hugegraph hstore to gh-dis-release

### PR DESCRIPTION
**Pull request by guoygang**

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java [hunk 1]
```diff
@@ -980,7 +980,7 @@ else if (ro.type().isGrantOrUser()) {
         if (!(actionPerm == HugePermission.READ && ro.type().isSchema()) &&
             auditLimiter.tryAcquire()) {
             String status = result == null ? "denied" : "allowed";
-            LOG.info("User '{}' is {} to {} {}", username, status, action, ro);
+        //    LOG.info("User '{}' is {} to {} {}", username, status, action, ro);
```
Comment (guoygang) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java]: Why remove user's permission access log

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 2]
```diff
@@ -742,7 +743,7 @@ private PageIds doIndexQueryOnce(IndexLabel indexLabel,
     }
 
     @Watched(prefix = "index")
-    private Set<MatchedIndex> collectMatchedIndexes(ConditionQuery query) {
+   protected Set<MatchedIndex> collectMatchedIndexes(ConditionQuery query) {
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 3]
```diff
@@ -1551,7 +1552,7 @@ protected void removeIndex(IndexLabel indexLabel) {
         this.doRemove(this.serializer.writeIndex(index));
     }
 
-    private static class MatchedIndex {
+   protected static class MatchedIndex {
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: align?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 4]
```diff
@@ -1641,30 +1651,34 @@ private void removeLeftIndexIfNeeded(Map<Id, HugeVertex> vertices) {
...
-            // Process results that query from left index or primary-key
-            if (query.resultType().isVertex() == elem.type().isVertex() &&
-                !rightResultFromIndexQuery(query, elem)) {
-                // Only index query will come here
-                return false;
-            }
-            return true;
-        });
+                return filterElement(query, elem);
+            });
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: align?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 5]
```diff
@@ -1641,30 +1651,34 @@ private void removeLeftIndexIfNeeded(Map<Id, HugeVertex> vertices) {
                                                 Iterator<T> results,
                                                 Query query) {
         // Filter unused or incorrect records
-        return new FilterIterator<T>(results, elem -> {
+            return new FilterIterator<T>(results, elem -> {
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 6]
```diff
@@ -60,7 +60,7 @@
 
     public SchemaElement(final HugeGraph graph, Id id, String name) {
         E.checkArgumentNotNull(id, "SchemaElement id can't be null");
-        E.checkArgumentNotNull(name, "SchemaElement name can't be null");
+//        E.checkArgumentNotNull(name, "SchemaElement name can't be null");
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java]: why remove check?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 7]
```diff
@@ -69,8 +69,8 @@ public SchemaElement(final HugeGraph graph, Id id, String name) {
     }
 
     public HugeGraph graph() {
-        E.checkState(this.graph != null,
-                     "Graph is null of schema '%s'", this.name);
+//        E.checkState(this.graph != null,
+//                     "Graph is null of schema '%s'", this.name);
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java]: why remove check?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeElement.java [hunk 8]
```diff
@@ -73,7 +73,7 @@
     private boolean defaultValueUpdated;
 
     public HugeElement(final HugeGraph graph) {
-        E.checkArgument(graph != null, "HugeElement graph can't be null");
+//        E.checkArgument(graph != null, "HugeElement graph can't be null");
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeElement.java:76]: Why remove check

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java [hunk 9]
```diff
@@ -314,7 +314,7 @@ private int save(Collection<HugeServerInfo> servers) {
             return this.dbExecutor.submit(callable).get();
         } catch (Throwable e) {
             throw new HugeException("Failed to update/query server info: %s",
-                                    e, e.toString());
+                                     e, e.toString());
```
Comment (guoygang) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java]: align?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java (hunk 2 above)
Comment (ssisme2019) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: done

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+import com.baidu.hugegraph.store.term.HgPair;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.baidu.hugegraph.store.client.util.HgStoreClientConst.ALL_PARTITION_OWNER;
+
+public class HstoreNodePartitionerImpl implements HgStoreNodePartitioner, HgStoreNodeProvider, HgStoreNodeNotifier {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    private HgStoreNodeManager nodeManager;
+
+    protected  HstoreNodePartitionerImpl(){
+
+    }
+    public HstoreNodePartitionerImpl(String pdPeers) {
+        pdClient = HstoreSessionsImpl.getDefaultPdClient();
+    }
+
+    public HstoreNodePartitionerImpl(HgStoreNodeManager nodeManager, String pdPeers) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        this(pdPeers);
+        this.nodeManager = nodeManager;
+    }
+
+    /**
+     * 查询分区信息，结果通过HgNodePartitionerBuilder返回
+     */
+    @Override
+    public int partition(HgNodePartitionerBuilder builder, String graphName, byte[] startKey, byte[] endKey) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        try {
+            if (HgStoreClientConst.ALL_PARTITION_OWNER == startKey) {
+                List<Metapb.Store> stores = pdClient.getActiveStores(graphName);
+                stores.forEach(e -> {
+                    builder.add(e.getId(), -1);
+                });
+            } else if (startKey == endKey) {
+                HgPair<Metapb.Partition, Metapb.Shard> partShard = pdClient.getPartition(graphName, startKey);
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        if (storeNotice.getPartitionIds() != null) {
+            storeNotice.getPartitionIds().forEach(partId -> {
+                pdClient.invalidPartitionCache(graphName, partId);
+            });
+        }
+        return 0;
+    }
+
+    public Metapb.GraphWorkMode setWorkMode(String graphName, Metapb.GraphWorkMode mode) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        // store列表
+        for (String address : hstorePeers.split(",")) {
+            storeMap.put((long) address.hashCode(), address);
+        }
+        // 分区列表
+        for (int i = 0; i < partitionCount; i++)
+            leaderMap.put(i, (long) storeMap.keySet().iterator().next());
+    }
+
+    public FakeHstoreNodePartitionerImpl(HgStoreNodeManager nodeManager, String peers) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+        this(peers);
+        this.nodeManager = nodeManager;
+
+    }
+
+
+    @Override
+    public int partition(HgNodePartitionerBuilder builder, String graphName, byte[] startKey, byte[] endKey) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+
+    @Override
+    public HgStoreNode apply(String graphName, Long nodeId) {
+        return nodeManager.getNodeBuilder().setNodeId(nodeId)
+                .setAddress(storeMap.get(nodeId)).build();
+    }
+
+    @Override
+    public int notice(String graphName, HgStoreNotice storeNotice) {
+        if (storeNotice.getPartitionLeaders() != null && storeNotice.getPartitionLeaders().size() > 0) {
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+            leaderMap.putAll(storeNotice.getPartitionLeaders());
+        }
+        return 0;
+    }
+    public static class NodePartitionerFactory {
+      public static HstoreNodePartitionerImpl getNodePartitioner(HugeConfig config, HgStoreNodeManager nodeManager){
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+          if ( config.get(HstoreOptions.PD_FAKE) ) // 无PD模式
+              return new FakeHstoreNodePartitionerImpl(nodeManager,
+                      config.get(HstoreOptions.HSTORE_PEERS));
+          else
+              return new HstoreNodePartitionerImpl(nodeManager,
+                                                   config.get(HstoreOptions.PD_PEERS)
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreNodePartitionerImpl.java]: align

File: hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreOptions.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+                    "pd.fake", "Enable the fake PD service.",
+                    disallowEmpty(), false);
+    public static final ConfigOption<String> HSTORE_PEERS = new ConfigOption<>(
+            "hstore.peers", "The addresses of store nodes, separated with commas.",
+            disallowEmpty(), "localhost:9080");
+    public static final ConfigOption<Integer> PARTITION_COUNT = new ConfigOption<>(
+            "hstore.partition_count", "Number of partitions, which PD controls partitions based on.",
+            disallowEmpty(), 0);
+    public static final ConfigOption<Integer> SHARD_COUNT = new ConfigOption<>(
+            "hstore.shard_count", "Number of copies,which PD controls partition copies based on.",
```
Comment (guoygang) [inline hugegraph-hstore/src/main/java/com/baidu/hugegraph/backend/store/hstore/HstoreOptions.java]: align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java (hunk 7 above)
Comment (ssisme2019) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java]: done

File: hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeElement.java (hunk 8 above)
Comment (ssisme2019) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/structure/HugeElement.java:76]: There is an operator sink scenario where graph would be null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java (hunk 9 above)
Comment (ssisme2019) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java]: done

**Review by yanjinbing — APPROVED** [review]

**Pull request merged**